### PR TITLE
[shardformer] shardformer support falcon

### PR DESCRIPTION
**Pull request by flybird11111**
[shardformer] shardformer support falcon

[shardformer] shardformer support falcon

## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4882 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

shardformer support falcon

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/falcon.py [hunk 1]
```diff
@@ -0,0 +1,774 @@
+
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:1]: Please update support matrix in shard former's README.md and documentations

File: colossalai/shardformer/modeling/falcon.py [hunk 2]
```diff
@@ -0,0 +1,774 @@
...
+        return outputs  # hidden_states, present, attentions
+    
+    return forward
+
+def get_falcon_flash_attention_forward():
+    try:
+        from xformers.ops import memory_efficient_attention as me_attention
+    except:
+        raise ImportError("Error: xformers module is not installed. Please install it to use flash attention.")
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:165]: Can Falcon run flash attention without installing xformers?

File: colossalai/shardformer/modeling/falcon.py [hunk 3]
```diff
@@ -0,0 +1,774 @@
...
+        logger = logging.get_logger(__name__)
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if past_key_values is None:
+            past_key_values = tuple([None] * len(self.h))
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:278]: past_key_values is not supported for Shardformer pipeline, please warn

File: colossalai/shardformer/modeling/falcon.py [hunk 4]
```diff
@@ -0,0 +1,774 @@
...
+        
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if output_attentions:
+            logger.warning_once("output_attentions=True is not supported for pipeline models at the moment.")
+            output_attentions = False
+        if output_hidden_states:
+            logger.warning_once("output_hidden_states=True is not supported for pipeline models at the moment.")
+            output_hidden_states = False
+
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:437]: please warn if `use_cache` is True/`past_key_values` is not None

File: colossalai/shardformer/modeling/falcon.py [hunk 5]
```diff
@@ -0,0 +1,774 @@
...
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if output_attentions:
+            logger.warning_once("output_attentions=True is not supported for pipeline models at the moment.")
+            output_attentions = False
+        if output_hidden_states:
+            logger.warning_once("output_hidden_states=True is not supported for pipeline models at the moment.")
+            output_hidden_states = False
+
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:523]: please warn if `use_cache` is True/`past_key_values` is not None

File: colossalai/shardformer/modeling/falcon.py [hunk 6]
```diff
@@ -0,0 +1,774 @@
...
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:635]: please warn if `use_cache` is True/`past_key_values` is not None

File: colossalai/shardformer/modeling/falcon.py [hunk 7]
```diff
@@ -0,0 +1,774 @@
...
```
Comment (Fridge003) [inline colossalai/shardformer/modeling/falcon.py:721]: please warn if `use_cache` is True/`past_key_values` is not None

File: tests/test_booster/test_plugin/test_gemini_plugin.py [hunk 8]
```diff
@@ -96,6 +96,7 @@ def check_gemini_plugin(subset: str, init_method: str = "none", early_stop: bool
             "transformers_sam",
             "transformers_vit",
             "transformers_gpt_double_heads",  # TODO check why does the model fail to run using Gemini
+            "transformers_falcon"
```
Comment (Fridge003) [inline tests/test_booster/test_plugin/test_gemini_plugin.py:99]: You can add a TODO here ("check why falcon fails to run Gemini")

File: colossalai/shardformer/modeling/falcon.py (hunk 2 above)
Comment (flybird11111) [reply colossalai/shardformer/modeling/falcon.py:165]: Models with alibi can only use xformers.

File: colossalai/shardformer/modeling/falcon.py (hunk 3 above)
Comment (flybird11111) [reply colossalai/shardformer/modeling/falcon.py:287]: updated

File: colossalai/shardformer/modeling/falcon.py (hunk 1 above)
Comment (flybird11111) [reply colossalai/shardformer/modeling/falcon.py:1]: updated

**Review by Fridge003 — APPROVED** [review]

**Pull request merged**